src/charm.py

```diff
@@ -1,233 +1,294 @@
+            OAUTH_SCOPES,
+            OAUTH_GRANT_TYPES,
+        )
 
     @property
-    def external_hostname(self):
-        """Return the DNS listing used for external connections."""
-        return self.config["external-hostname"] or self.app.name
-
-    def _require_nginx_route(self):
-        """Require nginx-route relation based on current configuration."""
-        require_nginx_route(
-            charm=self,
-            service_hostname=self.external_hostname,
-            service_name=self.app.name,
-            service_port=HTTP_PORT,
-            tls_secret_name=self.config["tls-secret-name"],
-            backend_protocol="HTTP",
+    def _oauth2_proxy_pebble_layer(self) -> Layer:
+        """OAuth2 Proxy pebble layer."""
+        proxy_vars = {
+            "HTTP_PROXY": "JUJU_CHARM_HTTP_PROXY",
+            "HTTPS_PROXY": "JUJU_CHARM_HTTPS_PROXY",
+            "NO_PROXY": "JUJU_CHARM_NO_PROXY",
+        }
+
+        context = {
+            "OAUTH2_PROXY_COOKIE_SECRET": self._cookie_encryption_key,
+        }
+
+        for key, env_var in proxy_vars.items():
+            value = os.environ.get(env_var)
+            if value:
+                context.update({key: value})
+
+        layer_config = {
+            "summary": "oauth2 proxy layer",
+            "services": {
+                WORKLOAD_SERVICE: {
+                    "summary": WORKLOAD_SERVICE,
+                    "command": f"/bin/oauth2-proxy --config {CONFIG_FILE_PATH}",
+                    "startup": "enabled",
+                    "override": "replace",
+                    "environment": context,
+                    "on-check-failure": {"up": "ignore"},
+                }
+            },
+            "checks": {
+                "up": {
+                    "override": "replace",
+                    "period": "10s",
+                    "http": {"url": f"http://localhost:{OAUTH2_PROXY_API_PORT}/ready"},
+                }
+            },
+        }
+        return Layer(layer_config)
+
+    def _validate_pebble_plan(self) -> bool:
+        """Validate pebble plan."""
+        try:
+            plan = self._container.get_plan().to_dict()
+            return bool(plan["services"][WORKLOAD_SERVICE]["on-check-failure"])
+        except (KeyError, pebble.ConnectionError):
+            return False
+
+    def _render_config_file(self) -> str:
+        """Render the OAuth2 Proxy configuration file."""
+        oauth_integration = False
+
+        if self.oauth.is_client_created():
+            oauth_provider_info = self.oauth.get_provider_info()
+            oauth_integration = True
+
+        with open("templates/oauth2-proxy.toml.j2", "r") as file:
+            template = Template(file.read())
+
+        rendered = template.render(
+            authenticated_emails_file=ACCESS_LIST_EMAILS_PATH if self.config["authenticated-emails-list"] else None,
+            client_id=oauth_provider_info.client_id if oauth_integration else "default",
```

Comment:
Client id, secret and email domains are mandatory to start the application, the reason I put a "default" value is to allow to run oauth2-proxy without blocking it due to lack of hydra (and consecutively, postgres and ingress) integrations.